deploy/runtools/run_farm.py

```diff
@@ -1,361 +1,424 @@
-
 from __future__ import annotations
 
 import re
 import logging
 import time
+import abc
 import os
 from datetime import timedelta
 from fabric.api import run, env, prefix, put, cd, warn_only, local, settings, hide # type: ignore
 from fabric.contrib.project import rsync_project # type: ignore
 from os.path import join as pjoin
+import pprint
+from collections import defaultdict
 
-from awstools.awstools import instances_sorted_by_avail_ip, get_run_instances_by_tag_type, get_private_ips_for_instances, launch_run_instances, wait_on_instance_launches, terminate_instances, get_instance_ids_for_instances
+from awstools.awstools import instances_sorted_by_avail_ip, get_run_instances_by_tag_type, get_private_ips_for_instances, launch_run_instances, wait_on_instance_launches, terminate_instances, get_instance_ids_for_instances, aws_resource_names, MockBoto3Instance
 from util.streamlogger import StreamLogger
+from util.inheritors import inheritors
+from runtools.run_farm_deploy_managers import InstanceDeployManager, EC2InstanceDeployManager
 
-from typing import Dict, Optional, List, Union, TYPE_CHECKING
+from typing import Any, Dict, Optional, List, Union, Set, Type, Tuple, TYPE_CHECKING
 if TYPE_CHECKING:
     from mypy_boto3_ec2.service_resource import Instance as EC2InstanceResource
     from runtools.firesim_topology_elements import FireSimSwitchNode, FireSimServerNode
+    from runtools.run_farm_deploy_managers import InstanceDeployManager
 
 rootLogger = logging.getLogger()
 
-def remote_kmsg(message: str) -> None:
-    """ This will let you write whatever is passed as message into the kernel
-    log of the remote machine.  Useful for figuring what the manager is doing
-    w.r.t output from kernel stuff on the remote node. """
-    commd = """echo '{}' | sudo tee /dev/kmsg""".format(message)
-    run(commd, shell=True)
-
-class MockBoto3Instance:
-    """ This is used for testing without actually launching instances. """
-
-    # don't use 0 unless you want stuff copied to your own instance.
-    base_ip: int = 1
-    ip_addr_int: int
-    private_ip_address: str
-
-    def __init__(self) -> None:
-        self.ip_addr_int = MockBoto3Instance.base_ip
-        MockBoto3Instance.base_ip += 1
-        self.private_ip_address = ".".join([str((self.ip_addr_int >> (8*x)) & 0xFF) for x in [3, 2, 1, 0]])
-
+class Inst(metaclass=abc.ABCMeta):
+    """Run farm hosts that can hold simulations or switches.
+
+    Attributes:
+        MAX_SWITCH_SLOTS_ALLOWED: max switch slots allowed (hardcoded)
+        switch_slots: switch node slots
+        _next_switch_port: next switch port to assign
+        MAX_SIM_SLOTS_ALLOWED: max simulations allowed. given by `config_runfarm.yaml`
+        sim_slots: simulation node slots
+        sim_dir: name of simulation directory on the run host
+        instance_deploy_manager: platform specific implementation
+        hostname: hostname of the instance (normally set by the `RunFarm` type)
```

Comment:
throughout this codebase, isn't this really `hostname_or_ip`?

Reply:
My mistake, I thought a hostname encompassed an IP address. Since they are separate things I'll update the function name.

Reply:
I like this since it matches the terminology of Fabric.